cpp/ql/src/Security/CWE/CWE-190/TaintedAllocationSize.ql

```diff
@@ -15,31 +15,73 @@ import cpp
 import semmle.code.cpp.security.TaintTracking
 import TaintedWithPath
 
-predicate taintedChild(Expr e, Expr tainted) {
-  (
-    isAllocationExpr(e)
-    or
-    any(MulExpr me | me.getAChild() instanceof SizeofOperator) = e
-  ) and
-  tainted = e.getAChild() and
-  tainted.getUnspecifiedType() instanceof IntegralType
+/**
+ * Holds if `e` is a simple arithmetic expression that might be used to
+ * compute the size of an allocation, and `f` is a child of `e`.
+ */
+predicate arithmeticChild(Expr e, Expr f) {
+  e.(AddExpr).getAChild() = f or
+  e.(MulExpr).getAChild() = f
+}
+
+/**
+ * Holds if `e` is an allocation with size expression `size`, or `e` is a call
+ * to a 'wrapper' function with that size, that returns an allocation.  `mid`
+ * is the allocation that's directly wrapped and `alloc` is the true allocation
+ * at the end of the chain.
+ */
+predicate allocExprOrIndirect(Expr e, Expr size, Expr mid, AllocationExpr alloc) {
+  isAllocationExpr(alloc) and
+  size = alloc.getSizeExpr() and
+  mid = alloc and
+  e = alloc
+  or
+  exists(Function f, int arg, Expr midSize, ReturnStmt rtn |
+    // indirect (wrapped) alloc call
+    f = e.(FunctionCall).getTarget() and
+    rtn.getEnclosingFunction() = f and
+    (
+      // return alloc
+      mid = rtn.getExpr()
+      or
+      // return variable assigned with alloc
+      exists(LocalScopeVariable v |
+        v = rtn.getExpr().(VariableAccess).getTarget() and
+        mid = v.getAnAssignedValue()
+      )
+    ) and
+    // size is passed through
+    allocExprOrIndirect(mid, midSize, _, alloc) and
+    arithmeticChild*(midSize, f.getParameter(arg).getAnAccess()) and
+    size = e.(FunctionCall).getArgument(arg)
+  )
+}
+
+/**
+ * Holds if `e` is an allocation or a call to a 'wrapper' function that
+ * returns an allocation, and it is not itself wrapped by another
+ * function.
+ */
+predicate allocExprOrIndirectRoot(Expr e, Expr size, Expr alloc) {
+  allocExprOrIndirect(e, size, _, alloc) and
+  not exists(Expr src | src != e and allocExprOrIndirect(src, _, e, alloc))
 }
 
 class TaintedAllocationSizeConfiguration extends TaintTrackingConfiguration {
-  override predicate isSink(Element tainted) { taintedChild(_, tainted) }
+  override predicate isSink(Element tainted) { allocExprOrIndirectRoot(_, tainted, _) }
 }
 
 predicate taintedAllocSize(
-  Expr e, Expr source, PathNode sourceNode, PathNode sinkNode, string taintCause
+  Expr source, Expr alloc, PathNode sourceNode, PathNode sinkNode, string taintCause
 ) {
   isUserInput(source, taintCause) and
   exists(Expr tainted |
-    taintedChild(e, tainted) and
+    allocExprOrIndirectRoot(alloc, tainted, _) and
```

Comment:
I think this has false negatives if there's multiple layers of wrappers that are used by non-library code - e.g. if there's a `my_calloc` that calls `my_malloc` that calls `malloc`, tainted sizes in direct invocations of `my_malloc` could be missed.

Reply:
This commit has been removed from the PR, but I'll come back to it at some point.  I think you're right, and it certainly deserves a test case.